sphinx_immaterial/custom_admonitions.py

```diff
@@ -0,0 +1,320 @@
+        else:
+            self.add_name(admonition_node)
+        if "collapsible" in self.options:
+            admonition_node["collapsible"] = self.options["collapsible"]
+        if "no-title" in self.options and "collapsible" in self.options:
+            logger.warning(
+                "title is needed for collapsible admonitions",
+                location=admonition_node,
+            )
+            del self.options["no-title"]  # force-disable option
+        textnodes, messages = self.state.inline_text(title_text, self.lineno)
+        if "no-title" not in self.options and title_text:
+            title = nodes.title(title_text, "", *textnodes)
+            self.set_source_info(title)
+            admonition_node += title
+        admonition_node += messages
+        admonition_node["classes"] += [nodes.make_id(self.name)]
+        self.state.nested_parse(self.content, self.content_offset, admonition_node)
+        return [admonition_node]
+
+
+def on_builder_inited(app: Sphinx):
+    """register the directives for the custom admonitions and build the CSS."""
+    config = app.config
+    custom_admonitions: List[CustomAdmonitionConfig] = getattr(
+        config, "sphinx_immaterial_custom_admonitions"
+    )
+    setattr(app.builder.env, "sphinx_immaterial_custom_icons", {})
+    user_admonitions = []
+    for config_val in custom_admonitions:
+
+        admonition = CustomAdmonitionConfig(**dict(config_val))
+
+        class UserAdmonition(CustomAdmonitionDirective):
+            default_title = admonition.title
+            name = admonition.name
+
+        app.add_directive(
+            name=admonition.name, cls=UserAdmonition, override=admonition.override
+        )
+
+        # set variables for CSS template to match HTML output from generated directives
+        admonition.name = nodes.make_id(admonition.name)
+        admonition.icon = load_svg_into_builder_env(app.builder, admonition.icon)
+        user_admonitions.append(admonition)
+    setattr(app.builder.env, "sphinx_immaterial_custom_admonitions", user_admonitions)
+
+
+def on_build_finished(app: Sphinx, exception: Optional[Exception]):
+    """generates the CSS for icons and admonitions, then appends that to the
+    doc project's CSS"""
+    if exception is not None or app.builder.name not in ("html", "dirhtml"):
+        return
+    custom_admonitions = getattr(
+        app.builder.env, "sphinx_immaterial_custom_admonitions"
+    )
+    custom_icons = getattr(app.builder.env, "sphinx_immaterial_custom_icons")
+    env = jinja2.Environment(
+        loader=jinja2.FileSystemLoader(str(PurePath(__file__).parent))
+    )
+    template = env.get_template("custom_admonitions.css")
+    generated = template.render(
+        icons=custom_icons,
+        admonitions=custom_admonitions,
+    )
+    static_output = Path(app.outdir) / "_static" / "stylesheets"
+    for css in static_output.glob("main.*.min.css"):
```

Comment:
Since we are modifying the CSS here, no purpose is served by including a hash in the name.  Instead, we could change `tools/build/` to leave the CSS under the original names `main.css` and `palette.css`, and not put them in `_static` to be copied automatically.  Then here we could produce a single combined CSS file (containing `main.css`, `palette.css` if `config.theme.palette` is set, and the custom admonitions), compute its hash, and copy it to the output directory.

Reply:
Well the map files for the CSS would be useless. These are really only helpful in theme development because the absolute paths in the map files are specific to the machine that ran `npm run build`.

I do like this idea though, I just didn't want to mess with the build script and the HTML template. I'd actually be happier if the CSS file was named "sphinx_immaterial_theme" instead of just "main", so users are still free to use the common name "main" for their extra CSS file (s).

Reply:
This was always going to the grey area for me. I think the `build-finished` event is triggered too late for changing what CSS files are copied. I think the [`html-collect-pages` event](https://www.sphinx-doc.org/en/master/extdev/appapi.html#event-html-collect-pages) may be more suitable to customizing the CSS files that will be copied to the output.

Additionally, _I think_ our custom builder's `init_css_files()` may need to be modified to temporarily exclude the theme's main & palette CSS bundles. Then they could be combined into a file that gets copied into the `outdir` path.

Where I get hazy is about the html templates because they currently look for main and palette bundles when generating each page. I didn't want to alter them, so I just let them behave normally and copied the generated CSS into the main.css that was copied.

Do you have a specific idea about which events to use for this request?

Reply:
> Well the map files for the CSS would be useless. These are really only helpful in theme development because the absolute paths in the map files are specific to the machine that ran `npm run build`.

I think at least in some cases the map file also can contain the full original source, so that it doesn't matter if the paths are valid locally.

> 
> I do like this idea though, I just didn't want to mess with the build script and the HTML template. I'd actually be happier if the CSS file was named "sphinx_immaterial_theme" instead of just "main", so users are still free to use the common name "main" for their extra CSS file(s).

Agreed, I think that makes sense.  In fact we could concatenate the user's extra css files as well, though.

The change to the html template would be pretty minimal --- we could just delete the existing tags that load the css and rely on the extra css loading variables.

Reply:
> we could concatenate the user's extra css files

If we do and the user's CSS has a syntactical error (like a missing `}` or `;`), then wouldn't it make that harder to diagnose when all the CSS cannot be loaded by the browser? I'm not sure how browsers handle erroneous CSS files (if the scope of the error is considered).

Reply:
We can only partially minify the user CSS (remove LFs).

Reply:
> This was always going to the grey area for me. I think the `build-finished` event is triggered too late for changing what CSS files are copied. I think the [`html-collect-pages` event](https://www.sphinx-doc.org/en/master/extdev/appapi.html#event-html-collect-pages) may be more suitable to customizing the CSS files that will be copied to the output.
> 
> Additionally, _I think_ our custom builder's `init_css_files()` may need to be modified to temporarily exclude the theme's main & palette CSS bundles. Then they could be combined into a file that gets copied into the `outdir` path.
> 
> Where I get hazy is about the html templates because they currently look for main and palette bundles when generating each page. I didn't want to alter them, so I just let them behave normally and copied the generated CSS into the main.css that was copied.

I think it would be a simple enough change that it likely won't cause significant problems later.
> 
> Do you have a specific idea about which events to use for this request? 

I think env-check-consistency

Reply:
> We can only partially minify the user CSS (remove LFs with any surrounding whitespace).

We could try using [csscompressor](https://github.com/sprymix/csscompressor).

Reply:
This is done, but I didn't combine the user defined CSS files because sphinx doesn't provide a convenient method like `remove_css_file()` (as opposed to `add_css_file()`). When I peeked under the hood for `add_css_file()`, it seems to add the file to the `app.registry` and the builder (if any).

My attempts at using a 3rd party solution to minify CSS failed for SVG data (embedded in `url()` definitions). So, I figured it was just easier to just let the user minify their docs' extra CSS and let Sphinx do its thing.

Reply:
Apparently the env-check-consistency event fires too early to generate the CSS. I found that the builder env var used to store aggregated icons' data is missing all inline icons (added to env by the `visit_si_icon()`).

I was able to produce desired behavior by only generating the CSS on build-finished event and calling `add_css_file()` on config-inited event. As of now, this didn't produce any warnings about copying non-existing assets.